File: 4.8-bc/go/stage-3-hexagonal/commerce/domain/order.go
// Package domain — Commerce Bounded Context.
package domain

import (
	"context"
	"errors"
	"time"

	"github.com/google/uuid"
)

type Order struct {
	ID         uuid.UUID
	UserID     uuid.UUID
	TotalCents int64
	Status     string
	CreatedAt  time.Time
}

type Repository interface {
	Create(ctx context.Context, o Order) error
}

var ErrInvalidOrder = errors.New("commerce.invalid_order")

// OrderPlaced — domain event. Billing and Notifications BCs підписуються.
type OrderPlaced struct {
	OrderID    uuid.UUID
	UserID     uuid.UUID
	TotalCents int64
	At         time.Time
}

func (OrderPlaced) Name() string { return "commerce.OrderPlaced" }

File: 4.8-bc/ts/stage-3-hexagonal/src/commerce/domain/order.ts
// Commerce Bounded Context, доменні типи й event.
import type { DomainEvent } from "../../shared/events/bus.js";

export interface Order {
  id: string;
  userId: string;
  totalCents: number;
  status: string;
  createdAt: Date;
}

// OrderPlaced — domain event. Billing і Notifications BC підписуються.
export class OrderPlaced implements DomainEvent {
  static readonly EVENT_NAME = "commerce.OrderPlaced";
  readonly name: string = OrderPlaced.EVENT_NAME;

  constructor(
    public readonly orderId: string,
    public readonly userId: string,
    public readonly totalCents: number,
    public readonly at: Date,
  ) {}
}
